### DISCUSSION TITLE
How to deal with Discord API v7 decommissioning

### DISCUSSION THREAD

**Discussion by super-cooper** [super-cooper/memebot#103] [Ideas]
Since [`discord.py` has been discontinued](https://gist.github.com/Rapptz/4a2f62751b9600a31a0d3c78100287f1), we are in need of migrating our API wrapper to a fork. 

Initially, it seemed like a good idea to wait for the dust to settle and for one fork to rise above the rest, but that has still not quite happened yet. 

Now, the time to act has been thrust upon us. As of April 30th, 2022, [v7 of the Discord API will be decommissioned](https://github.com/discord/discord-api-docs/discussions/4510). This is the version that discord.py depends on:

![image](https://user-images.githubusercontent.com/14908880/153915695-74d8850b-66ce-4207-95a2-1a88c23827f0.png)

This means that we have 2 choices to keep memebot alive after April 30th:
1. Migrate the API wrapper to a different one earlier than planned
2. Monkey patch `discord.py` to use a newer API, preferably v10, but v8 or v9 would suffice to simply buy us more time, as they are only being deprecated on April 30th. 

# Migrating to a new wrapper
We have been following the development of [`discord-interactions`](https://github.com/interactions-py/library) since development on discord.py ceased in August 2021. This library attempts to tightly integrate Discord's interactions into a python wrapper, and closely abides by Discord's evolving philosophy towards bots. They [recently announced](https://github.com/interactions-py/library/discussions/501) a partnership with [`pycord`](https://github.com/pycord-development/pycord), a fork of the original discord.py, which could help fill in any gaps not fulfilled by discord-interactions. It is unclear if we presently require the capabilities of pycord. 

Switching to `discord-interactions` would require us to finally migrate memebot to slash commands. The amount of work required to do this is unknown. However, we could try migrating the existing commands to `pycord` just to keep memebot alive, and then move to `discord-interactions` in the future, since we do eventually want to move to slash commands. This horizontal migration via pycord could be easier than going straight to interactions. 

# Monkey patching
It might be beneficial to attempt the monkey patch first just to see if its possible. If it is, we can keep this up our sleeves if all else fails. Monkey patching, _I think_, would be as simple as editing the `BASE` variables shown in the above screenshop at startup. 

# Discussion
Please weigh in with any thoughts you have. The future of memebot's existence depends on this 🥺

**Comment by super-cooper** [comment]
I think all avenues should be tried, just to ensure we have _a_ solution when April 30th rolls around. I think it is unnecessary to merge/deploy any solutions until we are either closer to that date, or we are positive we have the best solution we can think of. In order, 

1. Check if monkey patching works
2. Check if `pycord` works as a drop-in replacement for `discord.py`
3. Check how much work it would take to migrate to interactions

**Reply by schefm2** [reply]
Is `discord-interactions` the most “official” solution then? I am unfamiliar with coding bots for discord but slash commands sounds like that feature discord recently added, which has the autocomplete suggestions after a forward slash. I am evil so my suggestion would be to immediately go all or nothing with it.

**Reply by super-cooper** [reply]
Interactions are the ideal solution, which would be a migration to slash commands and all the other neat UI doohickeys. However, the only reason I want to try the simpler solutions first is because I am unsure how easy it will be to implement interactions by April 30th, given work schedules and motivation and all. That is, however, the solution I eventually want to arrive at.

**Comment by super-cooper** [comment]
I forgot to post in here, but development of `discord.py` has resumed (lol) so we have new avenues to consider, depending on the direction they decide to go in. Plus, the deadline for our current API version has been pushed back to 2023. See https://github.com/discord/discord-api-docs/discussions/4725 for more context.
